src/main/java/com/iridium/iridiumteams/listeners/PlayerMoveListener.java

```diff
@@ -19,37 +21,52 @@ public class PlayerMoveListener<T extends Team, U extends IridiumUser<T>> implem
 
     @EventHandler
     public void onPlayerMove(PlayerMoveEvent event) {
-        Player player = event.getPlayer();
+
         Location to = event.getTo();
-        Location from = event.getFrom();
         if (to == null) return; // This is possible apparently?
-        U user = iridiumTeams.getUserManager().getUser(player);
-        Optional<T> toTeam = iridiumTeams.getTeamManager().getTeamViaPlayerLocation(event.getPlayer(), to);
+
+        Location from = event.getFrom();
+
+        // might help speed things up - if the next location does not change blocks, why do anything?
+        // we don't need to check y, since we don't check anything against verticality.
+        if ((from.getBlockX() == to.getBlockX()) && (from.getZ() == to.getBlockZ())) return;
+
+        Player player = event.getPlayer();
+
         Optional<T> fromTeam = iridiumTeams.getTeamManager().getTeamViaPlayerLocation(event.getPlayer(), from);
+        Optional<T> toTeam = iridiumTeams.getTeamManager().getTeamViaPlayerLocation(event.getPlayer(), to);
+
         if (fromTeam.isPresent()) {
             iridiumTeams.getTeamManager().sendTeamTime(player);
             iridiumTeams.getTeamManager().sendTeamWeather(player);
         }
-        if (user.isFlying() && (to.getBlockX() != from.getBlockX() || to.getBlockZ() != from.getBlockZ()) && !user.canFly(iridiumTeams)) {
-            user.setFlying(false);
-            player.setAllowFlight(false);
-            player.setFlying(false);
-            player.sendMessage(StringUtils.color(iridiumTeams.getMessages().flightDisabled
-                    .replace("%prefix%", iridiumTeams.getConfiguration().prefix))
-            );
-        }
-        if (!toTeam.isPresent()) return;
-        if (!iridiumTeams.getTeamManager().canVisit(player, toTeam.get())) {
+
+        if (toTeam.isPresent() && !iridiumTeams.getTeamManager().canVisit(player, toTeam.get())) {
             event.setCancelled(true);
             player.sendMessage(StringUtils.color(iridiumTeams.getMessages().cannotVisit
                     .replace("%prefix%", iridiumTeams.getConfiguration().prefix))
             );
             return;
         }
 
+        // we should only be checking if the player is flying if the flight enhancement is enabled (this is a global config setting)
+        // we're not an anti-cheat, we don't care otherwise
+        U user = iridiumTeams.getUserManager().getUser(player);
+        if (iridiumTeams.getEnhancements().flightEnhancement.enabled && user.isFlying()) {
+            if (!user.canFly(iridiumTeams) && player.getGameMode() != GameMode.CREATIVE && player.getGameMode() != GameMode.SPECTATOR) {
```

Comment:
FWI, I moved this logic here, because we only want it to not disable their flight if they are in these gamemodes, having it in the canFly method will have a side effect of letting them use /is fly, which isnt a massive deal, but I think it makes more sence here